### PR TITLE
Add wgsl_version_at_least builtin function.

### PR DESCRIPTION
**Pull request by dneto0**
Define notion of language version with major and minor numbers.

State a backward compatility rule within the same version of WGSL.

This is WGSL version 0.0

Contributes to #3149

### CODE REVIEW

File: wgsl/index.bs
Comment (alan-baker) [inline wgsl/index.bs:11614]: This description seems wrong. This shader will run on an implementation that supports 1.0 or greater.

**Review by alan-baker — CHANGES_REQUESTED** [review]

File: wgsl/index.bs
Comment (dneto0) [reply wgsl/index.bs:9297]: Good question.  Hadn't thought of that.  I don't think so. But also I don't think anybody intends to do that.  Would be good to agree on that.

**Comment by litherum** [comment]
Feedback:

1. We're not super happy with this direction, but are willing to compromise.
2. The CG agreed that future versions of WGSL will be backwards compatible, so there's no need to have 2 version numbers (major and minor). We can get by with 1 (like how javascript just has "ES 2016" for example)
3. If the implementation is going to be exposing which version number it supports, it's necessary that the exposure happens at the API level (regardless of whether it's exposed in the shading language)
4. A requirement of this exposure is that, if a browser has implemented a feature from the next version of the language - but not _all_ the features in the next version, the browser can ship that feature if it's ready and spec-compatible. The browser wouldn't bump their reported version number in this case. Version numbers shouldn't hold up shipping new features for the sake of other unrelated features which happen to be in the same language version.

**Comment by dneto0** [comment]
> 4\. A requirement of this exposure is that, if a browser has implemented a feature from the next version of the language - but not _all_ the features in the next version, the browser can ship that feature if it's ready and spec-compatible. The browser wouldn't bump their reported version number in this case. Version numbers shouldn't hold up shipping new features for the sake of other unrelated features which happen to be in the same language version.

Yes, that is my intent and understanding. "Supporting a version" means "Supporting all the features in a stated version of the spec".    An implementation can support *more* things that may or may not appear in a later version than reported.

The normative text should express that concept since it's not clear.

**Comment by dneto0** [comment]
> Feedback:
> 
> 1. We're not super happy with this direction, but are willing to compromise.

Ack, and thank you.

> 2. The CG agreed that future versions of WGSL will be backwards compatible, so there's no need to have 2 version numbers (major and minor). We can get by with 1 (like how javascript just has "ES 2016" for example)

We're discussing at Google.

> 3. If the implementation is going to be exposing which version number it supports, it's necessary that the exposure happens at the API level (regardless of whether it's exposed in the shading language)

Agreed. I think this was agreed in the group meeting this week.

**Comment by dneto0** [comment]
Feedback from Mozilla in the group meeting this week included:

* It would be useful to actually get the number(s) rather than having to deduce it in a roundabout way like this.

**Comment by dneto0** [comment]
Regarding Mozilla's feedback:

> It would be useful to actually get the number(s) rather than having to deduce it in a roundabout way like this.

I should explain our thinking here.

We want to support a developer expressing clear intent:

* This shader may only run on an implementation supporting all of WGSL  version *BLAH*, or any later version.

This is useful because bodies of shaders are often developed (written,read,reviewed) far away from point-of-use (in the browser). 

We want to *increase the friction* for doing unfortunate things like:

         const_assert wgsl_version() == 1;

That would be too convenient to do with a builtin or constant that yielded **actual version numbers**.

**Comment by dneto0** [comment]
> > 4. A requirement of this exposure is that, if a browser has implemented a feature from the next version of the language - but not _all_ the features in the next version, the browser can ship that feature if it's ready and spec-compatible. The browser wouldn't bump their reported version number in this case. Version numbers shouldn't hold up shipping new features for the sake of other unrelated features which happen to be in the same language version.
> 
> Yes, that is my intent and understanding. "Supporting a version" means "Supporting all the features in a stated version of the spec". An implementation can support _more_ things that may or may not appear in a later version than reported.
> 
> The normative text should express that concept since it's not clear.

We had further discussion about this at Google. It's unclear this works well.

For concreteness:  Suppose features FOO and BAR appear first in v1.1, and the implementation currently implements all of v1.0 and reports v1.0.  (Still using major.minor nomenclature here, for exposition).

* CTS has positive tests (a feature works as described) and negative tests (things fail when they should)
* **How do you write CTS tests for FOO?** 
    * There is no way to say a particular test should pass for FOO but not for BAR.  
    * By design we don't turn it on via "enable FOO;", and there is no API mechanism for enabling FOO alone.

More broadly, how would one say that a browser is compliant with "the" spec, since FOO is not differentiated from BAR, in terms of declaring the shader's use of it, or requesting its presence from the API.


We'd like to probe why you'd want to implement **and expose** FOO independently of BAR.
Whatever the root concern there, perhaps we can address it with how "versions" of WGSL are formed: sequencing, granularity, and cadence.

**Sequencing:** what order features go into WGSL *core*.
**Granularity:**  put features together into a single core version when every implementor is ok implementing **and exposing** them at the same time.
**Cadence:**  how often do core versions get rolled out.

**Comment by litherum** [comment]
## Rationale

> We'd like to probe why you'd want to implement and expose FOO independently of BAR.

Good question!

First off, I'd like to state that we don't _want_ to intentionally deviate from the version progression of WGSL. We'd prefer to implement each version in its totality, monotonically.

However, if the spec for Foo is completely ready, (maybe even FOO is already shipped in other browsers), lots of customers want FOO, and we have implemented FOO, we want to be able to ship FOO without also having to implement a totally-unrelated BAR.

Some additional context (inside baseball): When we plan for upcoming releases, the unit of planning is "a feature." We always end up with way more features proposed than we have capacity to actually implement. So, each feature needs to be explained and justified. If the justification is "we need to implement this because the version number of WGSL says so," that's the weakest of weak sauce.

## Comparison with other web languages

For comparison, JavaScript, WebAssembly, CSS, and HTML are all compatible with this workflow:
- JavaScript's spec is a living standard, and [Test262](https://github.com/tc39/test262) follows it. Tests in Test262 have a "features" metadata, which allows individual UAs to enable/disable specific groups of tests, depending on which feature it's intended to test. If any browser naively runs Test262, it will report failures for features that are not implemented yet in that browser.
- WebAssembly has version numbers - it's now on version 2. It, too, is a living standard, and their [test suite](https://github.com/WebAssembly/spec/tree/main/test) follows it. WebAssembly is perhaps the closest comparison to WGSL, because they are an early-binding language too (cc @jimblandy). Considering WASM-SIMD, for example, there's no way in the WASM module to know whether it's supported or not, so the site author either consults caniuse, catches a compilation error, or does without.
- CSS and HTML are tested in [Web Platform Tests](https://github.com/web-platform-tests/wpt). Both of these languages are versioned in their own way, and that version is not exposed through software. As their specs are updated, WPT is updated accordingly. If any browser naively runs WPT, it will report failures for features that are not implemented yet in that browser.

## Full pointers

Regarding the "full pointers" example that was brought up in today's call, I'd propose we handle it as such:
- Create a `pointers-1.0.wgsl` test, guarded by an API check for `device.wgslVersionNumberIsAtLeast(1)`:
```
static assert versionNumberIsAtLeast(1);
// Exercise the set of things that are guaranteed to work in WGSL 1
let foo = &myLocalVariable;
```
- Create a `pointers-2.0.wgsl` test, guarded by an API check for `device.wgslVersionNumberIsAtLeast(2)`:
```
static assert versionNumberIsAtLeast(2);
// Exercise the set of things that are guaranteed to work in WGSL 2
let foo = &myThreadgroupVariable;
```
- Possibly: once all browsers are on WGSL 2, merge the two tests.

To @jimblandy's point today, nowhere in this proposal do we make sure that `&myThreadgroupVariable` fails on WGSL 1. The logic for that is: no one sits down at their computer and says to themself "boy gee golly I sure wish that my pointer wasn't allowed to point to a threadgroup variable."

For comparison, it's super common for the CSS Working Group to add a new value to an existing CSS property. When that happens, Web Platform Tests is (ideally) updated in tandem, and browsers that don't support the new value correctly fail the new test. Authors wanting to use the new value either test if it's there, or consult caniuse, or use Sass - they never consult _any_ version numbers.

However! It would be totally legit to write a negative test like:
```
let foo = "Strings shouldn't be supported.";
```
I believe we would have consensus in the WG that we don't expect strings to _ever_ be supported, and even if we ever _do_ start supporting strings, then at that time we could update the test. The test would then correctly fail on old browsers.

## Fears

It strikes me that the heart of this disagreement is just about granularity. Certainly, we all agree that the first browser to completely implement everything in version X should be able to ship at that point. (_Some_ browser needs to be first, of course.) This discussion is about whether the first browser to completely implement feature Y should be able to ship at _that_ point.

The key phrase above is "completely implement." That indicates not just that the code was typed into the computer, but also that the full standardization process was completed, and all the relevant stakeholders were given adequate chance to provide feedback and guidance. If a browser operates in good faith and incorporates advice from the community, following the full standards process, and finishes the feature, then of course that feature is ready to ship.

(For comparison, from the perspective of a browser not-yet shipping the feature, this is similar to the WebCodecs discussion we had maybe 6 months ago. Just because a browser isn't implementing Web Codecs at that time it was being discussed doesn't mean they're obviated from participating in the standards discussion. The feature is being standardized at that point, whether we want it to be or not, so it's up to every to participate in the discussion before it passes by.)

## Conclusion

Again, I'm not writing this post because I'm intentionally trying to thwart the versioning process or the standardization process. We want to abide by the features in WGSL versions! But we are trying to be realistic about shipping pressures and customer demand, in comparison to our own development capacity. I'm describing a kind-of escape hatch that I hope we'll never have to use.

**Comment by litherum** [comment]
Also: We don't have this problem for hardware-enabled features, because those are already necessarily scoped to just a single feature. So every browser can already ship those whenever the feature is fully standardized and implemented.

**Comment by dneto0** [comment]
Closing in favour of consensus at the 2022-02-16 F2F meeting.
See https://github.com/gpuweb/gpuweb/issues/3149#issuecomment-1433869176

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/gpu-web-f2f-2023-02-16-17#3792-wgsl_version_at_least">GPU Web F2F 2023-02-16/17</a></summary>

* KG: from the question about language evolution. What if you don't have restrictions on pointers any more? How do you test that?
* KG: Ben had a proposal last office hours.
* BC: talked with colleagues - not all unanimous. Recap: we originally pitched versions. WGSL 1.0 is what we agreed to. 1.1 - unrestricted pointer parameters. We'd proposed browsers have to implement everything in a version to advertise that version. Earlier versions - do not implement features from future versions. Bit contentious. To find middle ground, proposed feature names. "unrestricted_pointer_parameters". Browser can report it supports it. Discussed idea of WGSL versions, but a version's a collection of features bundled together. 1.1 to be unrestricted pointer parameters + a bunch of other things.
* BC: there's a danger in this - a browser could sit at 1.0 and cherry-pick forever. Many of us feel that's a disservice to our users. Good for browsers to go up in versions, don't need as many feature flags.
* BC: idea of letting browsers cherry-pick things if they want to, but have emphasis that browsers implement highest version they can, is the best compromise so far.
* KG: it's also something we can test. If you have this bucket of features, we can say you have all these available but only advertise the old language version - that's a bug.
* BC: builtin could be replaced by "requires" statement with version or list of feature flags. If browser doesn't support that, generates an error. "requires" doesn't change compiler behavior. Compilers don't have the 2^n permutation issue. They just report what they implement.
* MM: implementing features one-by-one isn't desirable either. Reason I wrote this - not to subvert the group. We want to move forward with everybody. Proposal of naming all features satisfies my list of desires. Will authors list 27 different things in their source code? If standards group is diligent about gathering up features and putting a version number on them, that'll be OK.
* MM: if we list feature names one-by-one - this is not in the rest of the web platform. We might as well do that for the web author. If you want to feature detect this, here's how to do it. In sum, generally positive.
* DS: other side - API can advertise it supports certain features.
* BC: Alan had concerns about spec being littered with features & feature dependencies. I don't have a great answer.
* KG: this form being proposed is almost exactly the same form as extensions in OpenGL. Doesn't get out of hand - there's a 2^n risk, but the features are mostly orthogonal and don't interact with each other that much. The CTS will have the worst of the challenges. E.g. I have while loops but not do loops.
* BC: few ways we'd try to prevent browsers having low base version but lots of features. A feature can require a base version.
* MM: thumbs down.
* KG: I understand the fear. But don't think we should worry about it. Take normal approach of standardization. Talk together, say: these are the group of things we want to put in the next version. Think we're find there.
* MM: let's tackle that problem if it occurs.
* BC: Yes. Was mostly for one feature overlapping another feature. If this feature depends on this set of other features -
* MM: I'm OK with technical reasons for dependencies. Ray tracing depends on bindless. Does it make sense that we force browser X to implement feature Y?
* KG: easier to define an extension when assuming presence of certain other extensions. Think ray tracing can require bindless, for example. If you need a subset of WGSL 1.1, you could say you require WGSL 1.1.
* MM: testable, too. Can't support ray tracing if you don't have bindless.
* RC: do we need an increasing version, or a skill tree?
* KG: a numbered version …
* MM: don't want to add a string type - function taking string returning bool.
* BC: it'd rather be a require directive.
* MM: OK, sounds better.
* BC: we were saying requires, then comma-separated list of things.
* MM: similar to enable directive. Work the same way.
* KG: different from "features" list?
* DS: do you need a different directive? Just use "enables"?
* DN: it's different. You shouldn't be able to use f16 if you haven't enabled the feature.
* BC: requires is a tiny option.
* CW: then we have navigator.doesSupportWGSLFeature?
* BC: think you should just list the features.
* MM: think this is just a FrozenArray on the device.
* KN: think it should be global on navigator.gpu.
* MM: / KN: these are all sugar.
* CW: maybe tuples are harder to implement on one backend or another. ;)
* KG: **consensus:** requires, something on the API side, enumeration of features.
* RC: should authors have to add "requires" to their shaders?
* KG: no. It's just if they want to make their shaders more robust.
* KG: one tradeoff, you don't get the WebGL style where you have to enable everything.
* BC: "requires" has a nice offline tooling story. Can ensure requires list matches the shader.
* KG: WebGL had a nice story for this - group should understand they're choosing a different direction and there will be more porting pains.
* RC: there are tradeoffs.
* 

</details>

**Pull request closed without merging**
